easybuild/tools/toolchain/compiler.py

```diff
@@ -277,19 +281,39 @@ def _set_optimal_architecture(self, default_optarch=None):
         :param default_optarch: default value to use for optarch, rather than using default value based on architecture
                                 (--optarch and --optarch=GENERIC still override this value)
         """
-        optarch = None
+        use_generic = False
+        optarch = build_option('optarch') 
         # --optarch is specified with flags to use
-        if build_option('optarch') is not None and build_option('optarch') != OPTARCH_GENERIC:
-            optarch = build_option('optarch')
-        # --optarch=GENERIC
-        elif build_option('optarch') == OPTARCH_GENERIC:
+        if optarch is not None:
+            # optarch has been parsed as a simple string
+            if isinstance(optarch, basestring):
+                if optarch == OPTARCH_GENERIC:
+                    use_generic = True
+
+            # optarch has been validated as complex string with multiple compilers and converted to a dictionary
+            elif isinstance(optarch, dict):
+                current_compiler = getattr(self, 'COMPILER_FAMILY', None)
+                if current_compiler in optarch:
+                    if optarch[current_compiler] == OPTARCH_GENERIC:
+                        use_generic = True
+                    else:
+                        optarch = optarch[current_compiler]
+                # no option for this compiler
+                else:
+                    optarch = None
+                    self.log.info("_set_optimal_architecture: no optarch found for compiler %s. Ignoring option.", 
+                            current_compiler)
+            else:
+                raise EasyBuildError("optarch is neither an string or a dict. This should never ever happen")
```

Comment:
@damianam we should show what's in `optarch` to make the error clearer? (although it's unlikely this will ever happen)